src/main/java/com/laytonsmith/core/functions/BasicLogic.java

```diff
@@ -147,6 +152,15 @@ public CBoolean exec(Target t, Environment env, Mixed... args) throws CancelComm
 			}
 		}
 
+		@Override
+		public CClassType getReturnType(Target t, List<CClassType> argTypes,
+				List<Target> argTargets, Environment env, Set<ConfigCompileException> exceptions) {
+			for(int i = 0; i < argTypes.size(); i++) {
+				StaticAnalysis.requireType(argTypes.get(i), Mixed.TYPE, argTargets.get(i), env, exceptions);
+			}
+			return CBoolean.TYPE;
+		}
```

Comment:
I kind of would rather invert this, and simply have the function return its types, and then the static analysis tool determines the correct signature and validates against that. So getReturnType would just return CBoolean.TYPE, and that's it in this case. Then a separate method would return the argument types.

This work was begun in another branch, NewCompiler, and there's several library classes for generating this information, which we can cherry pick and add to this branch. See https://github.com/EngineHub/CommandHelper/tree/NewCompiler/src/main/java/com/laytonsmith/core/arguments and see https://github.com/EngineHub/CommandHelper/blob/NewCompiler/src/main/java/com/laytonsmith/core/functions/ArrayHandling.java#L86 for an example use. 

For some functions, this is not sufficient, and we may very well need more complex logic like you have implemented here, and that's fine for a more complex function, but using this method for everything will end up with a TON of duplicated code, varying only slightly.

Reply:
I currently have two typecheck functions which can be overridden. One gets the AST terms, scope graph and anything it might need access to, and the default implementation gets the types from the arguments and passes them to the getReturnType() we see here. What we could do is change the default implementation of getReturnType() to do even lower level logic, which would indeed be fine for many simple functions. That way, there are three levels at which you can implement typechecking: Very little control, reasonable conditional typing control and full AST and scope graph access control.

Reply:
Yes, that seems reasonable. Actually, I think in almost all cases, we would want to implement this new method anyways, for documentation and reflective purposes, so we don't generally need an AST to figure out what arguments and return types are, so that probably would be required for all functions anyhow.

Reply:
The more I think about it, the more I think that's definitely true. We can use the overridable methods as a stop gap, but we do need to get to the point where we are formally and statically able to declare the signature of functions. And actually, to do that, we need to probably combine the return type and input parameters into one, so we can have overloaded signatures with varying return types based on parameters, so different than how I had done it in that other branch as well. Need to return a single Signature object which contains all signatures. Given the scope of this change, I think probably this should be kept out of scope for this PR, and we can revisit that once we get this all working. Good that you have validated that in general you're able to do this with the static analysis though.

Reply:
It would be great if documentation is generated from the same code that is used in typechecking, since then the documentation and implementation would always match. It'd be nice if you could complete that system and write a default getReturnType() implementation in AbstractFunction that calls that. I do see that you are using for example CInt.class where I use CInt.TYPE. I think it would be good to use CClassTypes everywhere to make all type handling in compile time work on the same system.

Reply:
Yes, CClassType was introduced after this branch. But user classes are not going to be representable through java classes, so this change was inevitable. But the principal is mostly the same.

Reply:
Considering that adding some signature to functions and using those to get both the return type and generate documentation can be done using the current implementation in the PR, and that it is out of scope for this PR, I'm marking the conversation as resolved here.